Add tests to CI

vadi2 opened the pull request.

vadi2 committed:
Create run-tests.yml

vadi2 committed:
Install dependencies

vadi2 committed:
Install apt dependencies

vadi2 committed:
Add missing cairo dep

vadi2 committed:
Add missing libpango1.0-dev dep

Photon89 committed:
Updated deps in CI control file

Photon89 committed:
Added missing space

Photon89 commented:
I updated the required dependencies according to the list in https://packages.ubuntu.com/jammy/shutter However, GitHub seems to be using Ubuntu 20.04 in CI which is missing two dependencies: libgoocanvas2-cairotypes-perl and libgtk3-imageview-perl 

We need to either wait till GitHub switches to 22.04 or install those by hand.

Photon89 committed:
Install the missing perl libs using cpanm

Photon89 committed:
Remove missing dependencies from the apt call

Photon89 committed:
Remove the correct dependencies from the apt call now

Photon89 committed:
Add libgoocanvas-2.0-dev as makedepend for GooCanvas2::CairoTypes

vadi2 commented:
Hmm. We could try running the tests in Docker then, but I haven't used that with actions myself.

Photon89 commented:
Maybe we can make it work the way it is, the only problem now is that the libs installed by cpanm are somewhere outside of @INC so perl doesn't find them. Probably they are somewhere in $HOME and this folder needs to be added to @INC. I am testing on an Ubuntu 20.04 VM and can reproduce the problem but don't yet know how to solve it...

DarthGandalf commented:
Add it to inc ? Either directly or via local::lib

ср, 2 мар. 2022 г., 12:11 Michael Kogan ***@***.***>:

> Maybe we can make it work the way it is, the only problem now is that the
> libs installed by cpanm are somewhere outside of @inc
> <https://github.com/inc> so perl doesn't find them. Probably they are
> somewhere in $HOME and this folder needs to be added to @inc
> <https://github.com/inc>. I am testing on an Ubuntu 20.04 VM and can
> reproduce the problem but don't yet know how to solve it...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/shutter-project/shutter/pull/442#issuecomment-1056859222>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACPLZGZMOVHB3R6M3RQD43U55LFLANCNFSM5KRHLZVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

vadi2 commented:
What is inc?

Photon89 commented:
Sorry, guys, my bad. The "--install-deps" option does what it tells to do, only installs the deps, not the actual module. I'm on it.

Photon89 committed:
Fixed the cpanm calls

Photon89 commented:
Well, now everything seems to be in place but some of the tests seem to fail because Shutter tries to do GUI stuff but we are in a CLI environment.

DarthGandalf commented:
Use xvfb

ср, 2 мар. 2022 г., 12:38 Michael Kogan ***@***.***>:

> Well, now everything seems to be in place but some of the tests seem to
> fail because Shutter tries to do GUI stuff but we are in a CLI environment.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/shutter-project/shutter/pull/442#issuecomment-1056884013>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACPLZFXWKCRBPBQGVW7TM3U55OL3ANCNFSM5KRHLZVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Photon89 committed:
Trying to enable GUI related tests 

Using xvfb as described in https://stackoverflow.com/questions/68292797/running-java-gui-tests-on-github-using-xvfb

Photon89 commented:
Looks like passing TEST_APP_SHUTTER_PATH=$(pwd) does not work any more when appended to the xvfb call, it should be prepended instead, but I don't see how to achieve this in the CI yaml file's syntax...

Photon89 committed:
Set environment variable outside of run call

Photon89 committed:
Try to fix the pwd call

Photon89 commented:
Looks like, I messed up so badly that the tests don't even try to run any more. :smile: Why do we need this TEST_APP_SHUTTER_PATH variable at all and cannot we just hardcode it for the CI test?

Photon89 committed:
Try to get the correct working directory

Photon89 commented:
Finally! :smile:

vadi2 commented:
Hot! Great work finishing this out. I can't approve, but you can 😃

Photon89 commented:
Thanks! I think, @ruzhnikov and @DarthGandalf should look at this, as they should know, how the tests should turn out.

DarthGandalf commented:
@vadi2 

> What is inc?

https://perldoc.perl.org/variables/@INC
https://metacpan.org/pod/local::lib

DarthGandalf commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,40 @@
+          libreadonly-perl librsvg2-common libsort-naturally-perl libwww-mechanize-perl libwww-perl \
+          libx11-protocol-other-perl libx11-protocol-perl libxml-simple-perl procps xdg-utils
+            
+          curl -L http://cpanmin.us | perl - --sudo App::cpanminus
```
This is dangerous, please don't curl http into execution, even without any encryption. Ubuntu has `cpanminus` package

Photon89 replied:
Thanks, fixed! For some reason it now fails to download two of the packages which it downloaded just fine previously...

Photon89 committed:
Install cpanm via package management

Photon89 committed:
Try to fix 404 errors on package installation

Photon89 commented:
Turned out, the package database got outdated after some hours, I inserted an `apt update` so it works now.

Ready to merge again!

Photon89 commented:
If there are no further comments on necessary changes, I'll merge the PR tomorrow.
